sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuAvroScan.scala

```diff
@@ -151,92 +151,276 @@ case class GpuAvroPartitionReaderFactory(
+   *                              reads per batch
+   * @return
+   */
+  protected def populateCurrentBlockChunk(
+      blockIter: BufferedIterator[BlockInfo],
+      maxReadBatchSizeRows: Int,
+      maxReadBatchSizeBytes: Long): Seq[BlockInfo] = {
+
+    val currentChunk = new ArrayBuffer[BlockInfo]
+    var numRows, numBytes, numAvroBytes: Long = 0
+
+    @tailrec
+    def readNextBatch(): Unit = {
+      if (blockIter.hasNext) {
+        val peekedRowGroup = blockIter.head
+        if (peekedRowGroup.count > Integer.MAX_VALUE) {
+          throw new UnsupportedOperationException("Too many rows in split")
+        }
+        if (numRows == 0 || numRows + peekedRowGroup.count <= maxReadBatchSizeRows) {
+          val estBytes = GpuBatchUtils.estimateGpuMemory(readDataSchema, peekedRowGroup.count)
+          if (numBytes == 0 || numBytes + estBytes <= maxReadBatchSizeBytes) {
+            currentChunk += blockIter.next()
+            numRows += currentChunk.last.count
+            numAvroBytes += currentChunk.last.blockDataSize
+            numBytes += estBytes
+            readNextBatch()
+          }
+        }
+      }
+    }
+
+    readNextBatch()
+    logDebug(s"Loaded $numRows rows from Avro. bytes read: $numAvroBytes. " +
+      s"Estimated GPU bytes: $numBytes")
+    currentChunk
+  }
+
+  /** Read a split into a host buffer, preparing for sending to GPU */
+  protected def readPartFile(
+      blockMeta: AvroBlockMeta,
+      partFilePath: Path,
+      conf: Configuration): (HostMemoryBuffer, Long) = {
+    withResource(new NvtxWithMetrics("Avro buffer file split", NvtxColor.YELLOW,
+        metrics("bufferTime"))) { _ =>
+      if (blockMeta.blocks.isEmpty) {
+        return (null, 0L)
+      }
+      val estOutSize = estimateOutputSize(blockMeta)
+      withResource(partFilePath.getFileSystem(conf).open(partFilePath)) { in =>
+        closeOnExcept(HostMemoryBuffer.allocate(estOutSize)) { hmb =>
+          withResource(new HostMemoryOutputStream(hmb)) { out =>
+            copyBlocks(blockMeta, in, out)
+            // check we didn't go over memory
+            if (out.getPos > estOutSize) {
+              throw new QueryExecutionException(s"Calculated buffer size $estOutSize is" +
+                s" too small, actual written: ${out.getPos}")
             }
-          })
-          AvroBlockMeta(reader.getHeader(), filteredBlocks)
+            (hmb, out.getPos)
+          }
         }
       }
-    } else {
-      AvroBlockMeta(new Header(), Seq.empty)
     }
   }
-}
 
-/**
- * Avro block meta info
- *
- * @param header the header of avro file
- * @param blocks the total block info of avro file
- */
-case class AvroBlockMeta(header: Header, blocks: Seq[BlockInfo])
+  /** Estimate the total size from the given block meta */
+  private def estimateOutputSize(blockMeta: AvroBlockMeta): Long = {
+    // For simplicity, we just copy the whole header of AVRO
```

Comment:
For now, it's ok to copy the whole header, but from a long-term point of view, We may need to figure out how to just copy the useful information.

Reply:
Yes, we may do it in the later PR.